### PR TITLE
Introduce logging extensibility, JSON formatter extension

### LINKED ISSUE #1847: Add JsonFormatter support for the log
Hi,

it would be gread to have possibility to configure the JsonFormatter for the logs.
We can overwrite the configuration with a @BuildItem and StartupEvent but this is not nice. 

For example:
```java
 void onStart(@Observes StartupEvent ev) {
        ArrayList<Handler> handlers = new ArrayList(2);
        Formatter formatter = new JsonFormatter();
        ConsoleHandler handler = new ConsoleHandler(formatter);
        handler.setLevel(Level.INFO);
        handlers.add(handler);
        InitialConfigurator.DELAYED_HANDLER.setHandlers((Handler[]) handlers.toArray(EmbeddedConfigurator.NO_HANDLERS));
        log.info("The application is starting...");
    }
```

Thanks
Andrej

### LINKED ISSUE #2409: Support custom logging handlers
It would be great if Quarkus supported registration of custom logging handlers (some frameworks like [Sentry](https://sentry.io) use it as the integration point by registering custom JUL handlers - see an example [here](https://github.com/getsentry/examples/blob/master/java/java.util.logging/src/main/resources/logging.properties))

### LINKED ISSUE #4666: Add tests for json and custom logging
As part of https://github.com/quarkusio/quarkus/pull/4121 we need to have more tests around json logging and custom logging handlers

### PR DESCRIPTION
**Pull request by ia3andy**
Continue #4121 Fixes #1847, fixes #2409, enables #3985, enables #3166, enables #453, replaces #2348, fixes #4666

### CODE REVIEW

File: extensions/logging-json/runtime/src/main/java/io/quarkus/logging/json/runtime/JsonConfig.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
      * Determine whether to enable the JSON console formatting extension, which disables "normal" console formatting.
      */
     @ConfigItem(name = ConfigItem.PARENT)
-    boolean enabled;
+    boolean enable;
```
Comment (dmlloyd) [inline extensions/logging-json/runtime/src/main/java/io/quarkus/logging/json/runtime/JsonConfig.java]: This is not part of the test so it should be either a separate commit on top of the original commit that introduced the class or else it should be squashed into that commit.

File: extensions/logging-json/runtime/src/main/java/io/quarkus/logging/json/runtime/LoggingJsonRecorder.java [hunk 2]
```diff
@@ -11,7 +11,7 @@
 @Recorder
 public class LoggingJsonRecorder {
     public RuntimeValue<Optional<Formatter>> initializeJsonLogging(final JsonConfig config) {
-        if (!config.enabled) {
+        if (!config.enable) {
```
Comment (dmlloyd) [inline extensions/logging-json/runtime/src/main/java/io/quarkus/logging/json/runtime/LoggingJsonRecorder.java]: Here as well

**Review by dmlloyd — APPROVED** [review]

**Comment by gsmet** [comment]
CI is failing:
```
[ERROR] io.quarkus.logging.JsonFormatterTest  Time elapsed: 0.009 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at sun.misc.Unsafe.ensureClassInitialized(Native Method)
	at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:156)
	at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1088)
	at java.lang.reflect.Field.getFieldAccessor(Field.java:1069)
	at java.lang.reflect.Field.get(Field.java:393)
	at org.junit.platform.commons.util.ReflectionUtils.lambda$tryToReadFieldValue$5(ReflectionUtils.java:621)
	at org.junit.platform.commons.function.Try.lambda$call$0(Try.java:57)
	at org.junit.platform.commons.function.Try.of(Try.java:93)
	at org.junit.platform.commons.function.Try.call(Try.java:57)
	at org.junit.platform.commons.util.ReflectionUtils.tryToReadFieldValue(ReflectionUtils.java:621)
	at org.junit.jupiter.engine.descriptor.ExtensionUtils.lambda$registerExtensionsFromFields$3(ExtensionUtils.java:109)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.jupiter.engine.descriptor.ExtensionUtils.registerExtensionsFromFields(ExtensionUtils.java:104)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:147)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:77)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.NullPointerException
	at java.util.Properties$LineReader.readLine(Properties.java:434)
	at java.util.Properties.load0(Properties.java:353)
	at java.util.Properties.load(Properties.java:341)
	at io.quarkus.test.QuarkusUnitTest.withConfigurationResource(QuarkusUnitTest.java:398)
	at io.quarkus.logging.JsonFormatterTest.<clinit>(JsonFormatterTest.java:25)
	... 44 more
```

File: docs/src/main/asciidoc/logging.adoc [hunk 3]
```diff
@@ -75,6 +75,59 @@ The logging format string supports the following symbols:
...
+of the Quarkus application is captured by a service which can, for example, process and store the log information for
+later analysis.
+
+[id="json_logging"]
+==== JSON Logging
+
+In order to configure JSON logging, the `quarkus-logging-json` extension may be employed.  Add this extension to your
+application POM as the following snippet illustrates.
+
+NOTE: This snippet assumes that your application POM is configured in the manner described in the
```
Comment (gsmet) [inline docs/src/main/asciidoc/logging.adoc]: This sentence is unfinished.

File: docs/src/main/asciidoc/logging.adoc [hunk 4]
```diff
@@ -75,6 +75,59 @@ The logging format string supports the following symbols:
...
+[cols="<m,<m,<2",options="header"]
+|===
+|Property Name|Default|Description
+|quarkus.log.json|true|Determine whether JSON logging is enabled; disabling this property will allow the normal console logging to function.
+|quarkus.log.json.pretty-print|false|Determine whether the JSON output is "pretty-printed".
+|quarkus.log.json.date-format|default|Specify the date format to use; `default` indicates that the default format is used.
+|quarkus.log.json.record-delimiter||Specify a record delimiter to add. By default, no delimiter is used.
+|quarkus.log.json.zone-id|default|Specify the time zone ID to use for logging dates; `default` indicates that the default is used.
+|quarkus.log.json.exception-output-type|detailed|Specify the exception output type; values of `detailed`, `formatted`, or `detailed-and-formatted` are allowed.
+|quarkus.log.json.print-details|false|Specify whether detailed caller information should be logged.
```
Comment (gsmet) [inline docs/src/main/asciidoc/logging.adoc]: You should have a generated config file that you can use.

They are in `target/asciidoc/generated/config/` at the root of the project and we now generate them for config groups too.

File: extensions/logging-json/deployment/src/main/java/io/quarkus/logging/json/deployment/LoggingJsonSteps.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package io.quarkus.logging.json.deployment;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.ExecutionTime;
+import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.LogConsoleFormatBuildItem;
+import io.quarkus.logging.json.runtime.JsonConfig;
+import io.quarkus.logging.json.runtime.LoggingJsonRecorder;
+
+public final class LoggingJsonSteps {
```
Comment (gsmet) [inline extensions/logging-json/deployment/src/main/java/io/quarkus/logging/json/deployment/LoggingJsonSteps.java]: Maybe call it `*Processor`? It's the generally used name. There are a few different ones but let's try to settle on one convention.

File: extensions/logging-json/deployment/pom.xml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-logging-json-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../</relativePath>
+    </parent>
+
+    <artifactId>quarkus-logging-json-deployment</artifactId>
+    <name>Quarkus - JSON Logging - Deployment</name>
```
Comment (gsmet) [inline extensions/logging-json/deployment/pom.xml]: It's minor but maybe: `Quarkus - Logging - JSON - Deployment` in case we have several logging components in the future?

Same for the parent and runtime poms obviously.

**Review by gsmet — DISMISSED** [review]
I added a few minor comments and suggestions here and there.

The most important one being the unfinished sentence in the doc.

**Comment by ia3andy** [comment]
> JsonFormatterTest

Oups, thanks CI, I commited a file by mistake, it should be ok now..

**Comment by ia3andy** [comment]
@dmlloyd the other comments are more on your juridiction.. could u have a look? I could do it if you give me the go..

File: extensions/logging-json/deployment/pom.xml (hunk 6 above)
Comment (dmlloyd) [reply extensions/logging-json/deployment/pom.xml]: Sure.

File: extensions/logging-json/deployment/src/main/java/io/quarkus/logging/json/deployment/LoggingJsonSteps.java (hunk 5 above)
Comment (dmlloyd) [reply extensions/logging-json/deployment/src/main/java/io/quarkus/logging/json/deployment/LoggingJsonSteps.java]: I never liked `Processor` because it doesn't really process anything; these classes are just collections of disparate (possibly even only casually related) build steps.  So using the convention of `Steps` makes the most sense as a convention in my mind.

File: docs/src/main/asciidoc/logging.adoc (hunk 4 above)
Comment (dmlloyd) [reply docs/src/main/asciidoc/logging.adoc]: That shows how old this PR actually is :)

File: docs/src/main/asciidoc/logging.adoc (hunk 3 above)
Comment (dmlloyd) [reply docs/src/main/asciidoc/logging.adoc]: Oops!

**Comment by ia3andy** [comment]
@gsmet can we merge this?

File: docs/src/main/asciidoc/logging.adoc (hunk 4 above)
Comment (ia3andy) [reply docs/src/main/asciidoc/logging.adoc]: As discussed, this will be done later in another PR

**Comment by jtama-op** [comment]
When is this extension going to be available ? This is most needed for applications deployed on cloud...

**Comment by gsmet** [comment]
It will be included in our next release. CR on Dec 9th, Final on Dec 16th.

**Comment by JoePullen** [comment]
Is there a reason that the possibility to provide a customer formatter is only available for console logging ? 
We are moving applications from wildfly and EAP which have a custom json formatter (included as a module and configured in standalone) specifically to produce json log file formats which are injested to elastic via kafka using filebeat and it would be great to be able to use all the file logging functions (such as rollover etc) in quarkus but with our existing custom formatter.

**Comment by dmlloyd** [comment]
There isn't a strong reason, no.  We do need to get #5794 worked out I think before it can be addressed though.  That is probably where the discussion should continue.
